Split state and view in SelectedItemList

Adjective-Object opened the pull request:
#### Pull request checklist

- ~~[ ] Addresses an existing issue: Fixes #0000~~
- [x]  Include a change request file using `$ npm run change`

#### Description of changes

Splits SelectedItemList into state + view components

This is to:
- avoid supporting a component transitioning between controlled / uncontrolled over its lifetime
- allow for reuse of the selectedItemList's state in UnifiedPickers w/o duplicating the logic.
- make clear when a consumer can expect which callbacks to be called (& how that relates to props)

I also took the opportunity to:
- move from `ref` to `componentRef` in the list (this simplifies things a lot in pickers for the time being)
  - Since FloatingSuggestions is a BaseComponent, it needs to be referred to by `componentRef`
  - Avoids having to refer to components with refs on them with `& React.RefAttributes<FooRefType>` everywhere. This also introduces a lot of unnamed types into error logs and complicates debugging
- rename old references to `FloatingSuggestions` from `Picker` -> `Suggestions`.



@joschect 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9360)

jetao-msft commented on packages/experiments/src/components/SelectedItemsList/SelectedItemsList.tsx:
```diff
@@ -1,131 +1,221 @@
+  props: IControlledSelectedItemListProps<T> | IUncontrolledSelectedItemListProps<T>
+): props is IControlledSelectedItemListProps<T> => (props as any).selectedItems !== undefined;
+
+export const SelectedItemsList = <TItem extends BaseSelectedItem>(props: ISelectedItemsListProps<TItem>) => {
```
Given the importance of the `selectedItems` prop, maybe include a longer description of what will happen if you pass that prop vs. not in the `SelectedItemsList.types.ts` docstrings. Something like "If you pass this prop, you are taking full responsibility for managing the state of the selection and all selection-changing callbacks"

jetao-msft commented on packages/experiments/src/components/SelectedItemsList/SelectedItemsList.tsx:
```diff
@@ -1,131 +1,221 @@
+ * Uses selection indeces as its own thing
+ * @param selection the selection to get the indeces of as state.
+ */
+const useSelectionIndeces = (selection: Selection): number[] => {
```
The memoized Selection object + methods seems to suggest to me that Selections are mutated in place? In that case this may run into trouble as the selection indices won't update when the selection changes?

Adjective-Object replied:
Selections are mutated in-place, and issue resize events against EventGroups. Normally EventGroups are maintained by the fabric BaseComponent class, which initialize and dispose the group over a component's lifetime.

Adjective-Object commented:
Looks like a new rule has hit master that enforces relative imports on files... I fixed up the examples in this PR to make it compliant. 

I also removed some tests which start failing on importing EditingItem.

It looks like an issue with the jest config in `experiments`. See:
```
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    C:\Users\mahuangh\Projects\office-ui-fabric-react\packages\utilities\lib\EventGroup.js:2
    import { assign } from './object';
           ^

    SyntaxError: Unexpected token {

      1 | import * as React from 'react';
      2 | import { Selection, SELECTION_CHANGE } from 'office-ui-fabric-react/lib/Selection';
    > 3 | import { EventGroup } from '@uifabric/utilities/lib/EventGroup';
        | ^

```

leddie24 commented:
Hi @Adjective-Object, is this still in progress? Just pinging to make sure this PR doesn't go stale.  Thank you!

Adjective-Object commented:
Nope, this is waiting on review. I talked to @joschet out of band but maybe
it slipped his plate?

On Mon, Jun 17, 2019, 13:48 Eddie Liu, <notifications@github.com> wrote:

> Hi @Adjective-Object <https://github.com/Adjective-Object>, is this still
> in progress? Just pinging to make sure this PR doesn't go stale. Thank you!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OfficeDev/office-ui-fabric-react/pull/9360?email_source=notifications&email_token=AAI62SQKBKM4MEYLATMCNW3P272BPA5CNFSM4HU52IW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX4M5MI#issuecomment-502845105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI62SREFSKXL6RBBJTJVXLP272BPANCNFSM4HU52IWQ>
> .
>

dzearing commented:
@joschect Can you review this please?

joschect commented on packages/experiments/src/components/SelectedItemsList/ControlledSelectedItemsList.tsx:
```diff
@@ -0,0 +1,70 @@
+    }
+  }, [props.selectedItems, props.getItemCopyText, selectedIndices]);
+
+  React.useImperativeHandle(
```
Since this one is meant to be controlled should this be moved out? I'm not quite sure if that's the right move.

Adjective-Object replied:
This is the right call, but `copyItemsInSelectionToClipboard` is something that should probably still exist on the Controlled component API we present to users.

This + the above is pushing me towards having a separate "View" component & user-presented "Control" component.

joschect commented on packages/experiments/src/components/SelectedItemsList/MaybeControlled.ts:
```diff
@@ -0,0 +1,19 @@
+export type MaybeControlled<TControlledProps, TUncontrolledProps> =
```
Maybe monad! :P

joschect commented on packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/SelectedPeopleList.tsx:
```diff
@@ -1,19 +1,52 @@
-);
+export interface IControlledSelectedPeopleListProps<TPersona>
+  extends PartiallyOptionalWithoutOnRenderItem<IControlledSelectedItemListProps<TPersona>> {}
+export interface IUncontrolledSelectedPeopleListProps<TPersona>
```
Nit I think some spaces in here might make this more readable.

Adjective-Object replied:
This is just run through `prettier`, I think?

Adjective-Object commented on packages/experiments/src/components/SelectedItemsList/ControlledSelectedItemsList.tsx:
```diff
@@ -0,0 +1,70 @@
+ * An externally-managed selected item list.
+ */
+export const ControlledSelectedItemsList = <TItem extends BaseSelectedItem>(props: IControlledSelectedItemListProps<TItem>) => {
+  const selection = React.useMemo(() => props.selection, [props.selection]);
```
This memo shouldn't be needed anymore, since this just references props.selection. In a prior revision, ControlledSelectedItemsList could manage selection itself.

Adjective-Object commented on packages/experiments/src/components/SelectedItemsList/ControlledSelectedItemsList.tsx:
```diff
@@ -0,0 +1,70 @@
+  React.useEffect(() => {
+    selection &&
+      selection.setItems(
+        props.selectedItems,
```
I'm of mixed opinion around this.

I originally put this here so someone consuming the component as controlled wouldn't have to manage the selection themselves. Within fabric as it is now, there's a standard where even when a component is 'controlled', smart objects like Selection still have their lifecycles managed by inner components.

I guess where it should live depends on whether or not "Selection" counts as part of the data model of a component.

Adjective-Object replied:
If it does, I see the argument for pulling this out of the Controlled component.

However, unless we wanted to force users to manage the selection object's lifecycle, we would have to present a different component to the user (e.g. one that only manages selection state and doesn't manipulate the actual data).

joschect replied:
I feel like the view shouldn't have an opinion of how the selection is given to it, there are other ways to manage selection that the consuming component might want to use, at least it seems like it might. The consumer doesn't necessarily need to use the Selection class right?

Adjective-Object commented:
Talked to @joschect  out of band about this -- In the interest of unblocking actually merging the Unified Pickers PR, we're going to go ahead with it as it is now and fix it up while it's still in the experiments package.

micahgodbolt commented:
@Adjective-Object is there an issue with this PR making it change 1000+ files? Oh, seems you are trying to merge this into the fabric-7 branch. Are you intending to merge this into master?

micahgodbolt commented:
Ok, switched to master, and its 26 files changed.

micahgodbolt commented:
This is 100% experiments update, and @joschect has already signed off on it. I removed all of the old reviewiers (from when this was merging into fabric-7 branch). Once this passes, it should be good to merge in.

dzearing commented:
Missing a change file. Can you run `npm run change` to add? Make sure to npm install latest bits if you haven't.

dzearing commented:
@Adjective-Object @joschect this seems to have fallen off the radar. What is left here?

dzearing commented:
Closing stale PRs. @Adjective-Object please feel free to reopen when you have the time. :)
